### PR TITLE
Core: Agentic observability for vitest and ghost stories

### PR DESCRIPTION
**Pull request by yannbf**
Closes #

## What I did

This PR does the following:
- Adds a telemetry utility to detect whether the run is at a specific first session window of a particular event
- Adds a Vitest reporter which uses that
- Improves error categorization
- Adds some cleanup about ghost stories

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

> [!CAUTION]
> This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-34537-sha-cd66a6a9`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-34537-sha-cd66a6a9 sandbox` or in an existing project with `npx storybook@0.0.0-pr-34537-sha-cd66a6a9 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-34537-sha-cd66a6a9`](https://npmjs.com/package/storybook/v/0.0.0-pr-34537-sha-cd66a6a9) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`sidnioulz/agentic-telemetry-ws2`](https://github.com/storybookjs/storybook/tree/sidnioulz/agentic-telemetry-ws2) |
| **Commit** | [`cd66a6a9`](https://github.com/storybookjs/storybook/commit/cd66a6a9da97e4db6a1f0ac6d0e12cba480b4101) |
| **Datetime** | Tue Apr 14 13:34:30 UTC 2026 (`1776173670`) |
| **Workflow run** | [24401970008](https://github.com/storybookjs/storybook/actions/runs/24401970008) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=34537`_
</details>

## Summary by CodeRabbit

* **New Features**
  * Agent-aware telemetry for test runs with aggregated metrics, durations, and session-aware emissions
  * Story-level test conversion and analysis utilities

* **Improvements**
  * More detailed test-run analysis (success rates, empty-render accounting, unique error counts)
  * Broader render/hook error detection and categorization
  * Render-analysis enabled in more scenarios; reporter state resets between runs
  * CLI event-log: include/exclude filtering and metadata masking

* **Tests**
  * New and expanded tests for analysis, telemetry, conversion, and error-categorization logic

### CODE REVIEW

File: code/addons/a11y/src/preview.tsx [hunk 1]
```diff
@@ -20,10 +20,10 @@ export const afterEach: AfterEach<any> = async ({
 }) => {
   const a11yParameter: A11yParameters | undefined = parameters.a11y;
   const a11yGlobals = globals.a11y;
-  const isGhostStories = !!globals.ghostStories;
+  const isInternalRenderAnalysis = !!globals.renderAnalysis;
```
Comment (yannbf) [inline code/addons/a11y/src/preview.tsx]: leave old logic for ghost stories only. let ai do the a11y test

File: code/core/src/telemetry/types.ts [hunk 2]
```diff
@@ -48,7 +48,8 @@ export type EventType =
   | 'ai-prepare'
   | 'ai-prepare-evidence'
   | 'ai-prepare-story-scoring'
-  | 'ai-prompt-nudge';
+  | 'ai-prompt-nudge'
+  | 'agent-test-run';
```
Comment (yannbf) [inline code/core/src/telemetry/types.ts]: ai-setup-selfheal-scoring

File: code/core/src/telemetry/types.ts [hunk 3]
```diff
@@ -48,7 +48,8 @@ export type EventType =
   | 'ai-prepare'
   | 'ai-prepare-evidence'
   | 'ai-prepare-story-scoring'
```
Comment (yannbf) [inline code/core/src/telemetry/types.ts]: ai-setup-final-story-scoring

File: code/core/src/shared/utils/categorize-render-errors.test.ts [hunk 4]
```diff
@@ -192,6 +213,24 @@ describe('categorize-render-errors', () => {
...
+        it('should categorize context not found errors', () => {
+          expect(categorizeError('context not found').category).toBe(
+            ERROR_CATEGORIES.MISSING_PROVIDER
+          );
+
+          expect(categorizeError('No provider found for context').category).toBe(
+            ERROR_CATEGORIES.MISSING_PROVIDER
+          );
+
+          expect(categorizeError('Component must be used without a provider').category).toBe(
```
Comment (Sidnioulz) [inline code/core/src/shared/utils/categorize-render-errors.test.ts]: If I understand correctly, it's saying there SHOULDN'T be a provider, not that there should?

Should there be a different category for this one?

File: code/core/src/shared/utils/analyze-test-results.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  >();
+
+  const uniqueErrorMessages = new Set<string>();
+
+  for (const r of failed) {
+    const { category, matchedDependencies } = categorizeError(r.error!, r.stack);
+
+    if (!map.has(category)) {
+      map.set(category, { count: 0, uniqueErrors: new Set(), matchedDependencies: new Set() });
+    }
```
Comment (Sidnioulz) [inline code/core/src/shared/utils/analyze-test-results.ts:30]: Could this be a `getOrInsert`?

File: code/core/src/core-server/utils/ghost-stories/parse-vitest-report.test.ts [hunk 6]
```diff
@@ -94,9 +94,9 @@ describe('parse-vitest-report', () => {
     it('should categorize errors and include them in the summary', () => {
       const mockVitestResults = {
         success: false,
-        numTotalTests: 4,
+        numTotalTests: 5,
         numPassedTests: 1,
-        numFailedTests: 3,
+        numFailedTests: 4,
```
Comment (Sidnioulz) [inline code/core/src/core-server/utils/ghost-stories/parse-vitest-report.test.ts:99]: Sanity check: any reason why the total changed?

File: code/addons/vitest/src/vitest-plugin/index.ts [hunk 7]
```diff
@@ -455,6 +473,21 @@ export const storybookTest = async (options?: UserOptions): Promise<Plugin[]> =>
         },
         { configDir: finalOptions.configDir }
       );
+
+      if (isTelemetryModuleEnabled()) {
```
Comment (Sidnioulz) [inline code/addons/vitest/src/vitest-plugin/index.ts:477]: Here we'll need to wrap the whole code into `telemetry` and pass it as a callback.

**Review by Sidnioulz — APPROVED** [review]
We need to update the disabled telemetry detection, and see comment on error categorisation. All the rest LGTM!
